### ISSUE TITLE
MTOON shader is not receiving shadows

### ISSUE THREAD

**Issue by alacamda** [V-Sekai/godot-vrm#110]
MTOON shader is not receiving shadows from other objects blocking light

**Comment by lyuma** [comment]
Hi @alacamda, thanks for opening an issue. Could you please provide some screenshots and reproduction steps and/or test scenes that reproduce this?

note that MToon is not a physically accurate rendering model, so the shadows will not function the same as with a conventional PBR shader.

**Comment by alacamda** [comment]
Here you can see a very simple test scene. In the first image, VRM character with Mtoon is not receiving any shadows from the wall. As you can see the shadows affect the car next to the character

![image](https://github.com/user-attachments/assets/c7b33e6d-db24-4926-a9e1-e2576b5699e1)

In the second image, the character is out of the zone of influence of the shadow projected by the wall, but it looks the same as it looks in the first picture.

![image](https://github.com/user-attachments/assets/676ac2cd-634a-4d96-a081-b1e632460c85)


It is easy to reproduce, just create a simple scene with a directional light and a big cube projecting a shadow. The Shadow projected by the cube darkens areas of 3d meshes like the car, but not the VRM character that does not look darker

Hope this helps!

**Comment by alacamda** [comment]
Thank you @lyuma for the quick reply!

I just tested with the Alicia model and I am getting the same effect. As you can see both Alicia's models look the same. I was expecting the Alicia model in the shadow (the one next to the car) to look darker (as you can see, the front part of the looks very dark as it is receiving shadow the cube wall). However, both Alicias look the same, they are not receiving any shadow from the cube wall.
![image](https://github.com/user-attachments/assets/1ac72840-abfa-46c6-a02a-0e75ca6cd78f)
This is with only one directional light in the scene. 

Just to be clear, if I adjust the directional light to be darker (by reducing the energy) then of course the whole scene looks darker, including Alicia model (as you can see in the screenshot below), but still it's not receiving any shadow coming from the cube. 

![image](https://github.com/user-attachments/assets/622680cf-8ff6-4974-9fe1-16bdf0e741ae)

I tried also changing the shade color of Alicia to black, but then it appears always black. What I am looking for is for Alicia to appear darker only when affected by the shadow projected by another mesh (in this case, the wall cube)

Hope I explained this in a clear way, sorry but English is not my native language!

Thank you again for your support!

**Comment by lyuma** [comment]
Okay I think I have a better idea of what you're looking for.

There are two things to try playing with. One is what is called the "Shadow Receive" rate. I realize that this slider defaults to 0.5, but maybe relatively ineffectual by default due to the sRGB curve (it will be effectively squared, so it really is closer to 0.25 rate). I would try moving this to 1.0 on your materials.

The other change you could try is to modify the shader as shown in the screenshot. Change the `diffuse_output =` line in the `LIGHT_IS_DIRECTIONAL` case to add a multiplication with `lightAttentuation` - This will allow it to go all the way down to black as in the PBR case, but just for shadows (since ATTENUATION in directional lights is only affected by shadowing).

![godette_shadowing](https://github.com/user-attachments/assets/2ff8972e-ec04-4456-821b-f8024f821a33)

I believe this is mostly what you want, but beware of self-shadowing as it may look odd. You can disable self-shadowing by setting your meshes to have "Cast Shadows" off, and then you may be able to play some layering tricks to still allow the VRM meshes to cast shadows on the ground, but I'm not so sure if this is actually possible. Anyway, here's how it could look with cast shadows off:
![godette_disable_self_shadowing](https://github.com/user-attachments/assets/b4c36d19-4aee-4565-b71a-980d8582a1b4)

Note that to accentuate the effect, I modified the WorldEnvironment to adjust sky contribution to 0. If you have sky contribution set higher or other GI, the shadows won't be pitch black but should include the ambient light, even with the above shader change. This is correct and should match PBR materials.

Hope this helps.

I don't currently plan to include the extra multiplication as a feature in MToon since it doesn't conform with the MToon spec, but if there is interest by more people, I could consider adding this feature.

**Comment by alacamda** [comment]
Yes! That's exactly what I was looking for! Sorry to log it as a bug, I checked the shader code before opening the buyg but couldn't find the solution! Thank you so much!
